lib/maths/CBoostedTreeImpl.cc

```diff
@@ -1445,5 +1451,50 @@ const double CBoostedTreeImpl::INF{std::numeric_limits<double>::max()};
 const CBoostedTreeHyperparameters& CBoostedTreeImpl::bestHyperparameters() const {
     return m_BestHyperparameters;
 }
+
+void CBoostedTreeImpl::computeShapValues(core::CDataFrame& frame, const TProgressCallback&) {
+    if (m_TopShapValues > 0) {
+        if (m_BestForestTestLoss == INF) {
+            HANDLE_FATAL(<< "Internal error: no model available for prediction. "
+                         << "Please report this problem.");
+            return;
+        }
+        bool successful;
+        auto treeFeatureImportance = std::make_unique<CTreeShapFeatureImportance>(
+            m_BestForest, m_NumberThreads);
+        TDoubleVec shapTotal;
+        std::size_t numberInputFields = m_Encoder->numberEncodedColumns() - 1;
+        // resize data frame to write SHAP values
+        std::size_t offset{frame.numberColumns()};
+        frame.resizeColumns(m_NumberThreads, frame.numberColumns() + numberInputFields);
+        TStrVec columnNames(frame.columnNames());
+        for (std::size_t i = 0; i < numberInputFields; ++i) {
+            columnNames[offset + i] = CDataFrameRegressionModel::SHAP_PREFIX +
+                                      frame.columnNames()[i];
+        }
+        frame.columnNames(columnNames);
+
+        std::size_t topShapValues = (m_TopShapValues < numberInputFields)
+                                        ? m_TopShapValues
+                                        : numberInputFields;
+        shapTotal = treeFeatureImportance->shap(frame, *m_Encoder, numberInputFields, offset);
+
+        // get indices of the top elements
+        TSizeVec indices(shapTotal.size());
+        std::iota(indices.begin(), indices.end(), offset);
+        std::nth_element(indices.begin(), indices.end(),
+                         indices.begin() + topShapValues - 1,
+                         [&shapTotal](std::size_t a, std::size_t b) {
+                             return shapTotal[a] > shapTotal[b];
+                         });
+        m_TopShapIndices = TSizeVec(topShapValues);
+        std::copy(indices.begin(), indices.begin() + topShapValues,
+                  m_TopShapIndices.get().begin());
```

Comment:
I'm reading this as we output top overall shap values for each feature for each example. It seems to me like it would be more useful to return separate result: feature shap values, and then top (local) shap values for each example. If we return indices for all shap values then the writer can simply iterate over them and select the top n. You'd then introduce fields called `shap.<feature foo>`, so different results would have different fields set. WDYT?

Reply:
I adjusted the code to output top shap values per row.